eth2near/eth2near-block-relay-rs/src/eth2near_relay.rs

```diff
@@ -140,60 +149,130 @@ impl Eth2NearRelay {
+    // `1` -- expected gap between attested block slot and signature slot
+    fn get_gap_between_finalized_and_signature_slot(update_submission_frequency: u64) -> u64 {
+        update_submission_frequency * ONE_EPOCH_IN_SLOTS + 2 * ONE_EPOCH_IN_SLOTS + 1
+    }
+
+    fn find_left_non_error_slot(&self, slot_lft: u64, slot_rgt: u64) -> (u64, bool) {
+        let mut slot = slot_lft;
+        while slot < slot_rgt {
+            match self.block_known_on_near(slot) {
+                Ok(v) => { return (slot, v) },
+                Err(_) => slot += 1,
+            };
+        }
 
-        Ok(false)
+        return (slot, false);
     }
 
-    fn search_slot_forward(&self, slot: u64) -> Result<u64, Box<dyn Error>> {
+    fn search_slot_forward(&self, slot: u64, max_slot: u64) -> Result<u64, Box<dyn Error>> {
         let mut current_step = 1;
-        while self.block_known_on_near(slot + current_step)? {
-            current_step *= 2;
+        let mut prev_slot = slot;
+        while slot + current_step < max_slot {
+            match self.block_known_on_near(slot + current_step) {
+                Ok(true) =>  {
+                    prev_slot = slot + current_step;
+                    current_step = min(current_step * 2, max_slot - slot);
+                },
+                Ok(false) => break,
+                Err(_) => {
+                    let (slot_id, slot_on_near) = self.find_left_non_error_slot(slot + current_step, max_slot);
+                    if slot_on_near {
+                        prev_slot = slot_id;
+                        current_step = min(current_step * 2, max_slot - slot);
+                    } else {
+                        current_step = slot_id - slot;
+                        break;
+                    }
+                }
+            }
         }
 
-        self.search_slot_backward(slot, slot + current_step)
+        self.binsearch_slot_range(prev_slot, slot + current_step)
     }
 
-    fn search_slot_backward(&self, start_slot: u64, last_slot: u64) -> Result<u64, Box<dyn Error>> {
+    fn binsearch_slot_range(&self, start_slot: u64, last_slot: u64) -> Result<u64, Box<dyn Error>> {
         let mut start_slot = start_slot;
         let mut last_slot = last_slot;
         while start_slot + 1 < last_slot {
             let mid_slot = start_slot + (last_slot - start_slot) / 2;
-            if self.block_known_on_near(mid_slot)? {
-                start_slot = mid_slot;
-            } else {
-                last_slot = mid_slot;
+            match self.block_known_on_near(mid_slot) {
+                Ok(true) => start_slot = mid_slot,
+                Ok(false) => last_slot = mid_slot,
+                Err(_) => {
+                    let (lft_slot, val) = self.find_left_non_error_slot(mid_slot, last_slot);
+                    if val == true {
+                        start_slot = lft_slot;
+                    } else {
+                        last_slot = mid_slot;
+                    }
+                }
```

Comment:
```suggestion
                Err(_) => {
                    let (left_slot, value) = self.find_left_non_error_slot(mid_slot, last_slot);
                    if value {
                        start_slot = left_slot;
                    } else {
                        last_slot = mid_slot;
                    }
                }
```

Reply:
fixed: https://github.com/aurora-is-near/rainbow-bridge/pull/775/commits/f4a259bbeff94d1126a2eb0a87760434c1836cdc